### PR TITLE
[firefox-webext-browser] Narrowed message and storage types

### PR DESCRIPTION
**Pull request by leaumar**
Authors: @jsmnbom 

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

Disclaimer: I tried following the testing/compiling steps from the readme but the test file already seems to not work as-is, and `index.d.ts` already contains a lint/compile violation on the `eval` function?

Since the various functions around local/sync storage and background/tabs messaging can _only_ return jsonified objects, I believe they should be typed as such. This prevents you from writing code that e.g. mistakenly assumes a field to be of type Date or some such. I've also added the "smarts" to the storage functions that make them properly predict the structure of the loaded object, which is always derived from the `keys` passed in.
TLDR: `any` is evil.

I'm not 100% sure about some of the `any`s I've changed to `JsonObject`  if they shouldn't also be ` | undefined`, I'll have to read the specs to get a better idea.

I'm putting this PR up half-finished as a starting point, POC, and basis for discussion. I'm lacking some insider knowledge/experience on DefinitelyTyped and this package to be able to get this PR to 100% done, but I assume you can help with that and do the parts you can more easily do yourself than explain to me? :)

### CODE REVIEW

**Review by jsmnbom — CHANGES_REQUESTED** [review]
Thanks a lot for taking the time to make this PR :) The thing is tho, that many of these api's don't only support JSON. they pretty much support anything you can throw at them, as long as they can be [structured cloned](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm). Now i do believe we could do some changes to make some of these api's types better. In an extension i worked on a bit ago i wrote this code, that allows for much better typing of the storage get and set. But this code should definitely be added at https://github.com/jsmnbom/definitelytyped-firefox-webext-browser so future automatic updates (when mozilla changes the schemas) don't overwrite the new code here.

**Comment by leaumar** [comment]
Are you sure there are cases where the _receiving_ end receives more than just json values though? One of my addons actually had that as a bug: my type assumed a field to be a `Date` (since that's what was put _in_), but the value loaded back from firefox's (or chrome's, can't remember) storage was a string instead. I'll check again later today, see what happens to a storage-loaded Date or RegExp...

**Comment by jsmnbom** [comment]
Hmm , it's possible that storage is always JSON, in which case browser.extensionTypes.PlainJSONValue should probably be used as the type. But once again, preferably not here, but in https://github.com/jsmnbom/definitelytyped-firefox-webext-browser.
MDN says that sendMessage can take any data that can be structurally cloned, but in onMessage it says it's json-ifiable, so MDN doesn't even agree with itself here.. so some testing is definitly needed.

**Comment by leaumar** [comment]
Huh. Seems the two browsers disagree quite a bit. Firefox restores the runtime object type, while chrome sets the bar at a lower, more predictable point. Kind of explains some of the bugs I've had over the years too and why I decided on json-in-json-out in my addons.

image
image

I guess if this types package is specifically meant for firefox, the only part of my suggestion that still makes sense is the key-based structure inferral. I don't think I'll put work into a PR on that project though, sorry. The essential code you need for that (assuming you can work directly in ts types, it wasn't possible with a competing types lib) is the following:
```typescript
type JsonOfString<Key extends string> = Record<Key, JsonValue>;
type JsonOfStrings<Keys extends string[]> = Record<Keys[number], JsonValue>;
type JsonOfObject<T extends Record<string | number | symbol, unknown>> = Record<keyof T, JsonValue>;

function get<Keys extends string | string[] | { [key: string]: unknown }>(keys?: Keys): Promise<
        Keys extends string
            ? JsonOfString<Keys>
            : Keys extends string[]
            ? JsonOfStrings<Keys>
            : Keys extends {[key: string]: unknown}
            ? JsonOfObject<Keys>
            : unknown
    >;
```
